rfd/0164-scoped-rbac.md

```diff
@@ -0,0 +1,956 @@
+
+#### Scoped Audit Log Events and Session Recordings
+
+The audit events and session recordings generated by activity in some scope, will have a property that binds them to the same scope. 
+
+This will let us filter and grant access to a subset of events and session recordings within a scope.
+
+#### Resources that can’t be created at non-cluster scopes
+
+Some resources don't have a clear cut behavior at certain scopes, like SSO connectors, or are difficult to define, like users. 
+
+That’s why we will prohibit creating those resources by roles granted by any scope other than the root. 
+
+Here is a list of resources that can’t be created at scopes:
+
+* SSO connectors
+* Users
+* Clusters
+* Login rules
+* Devices
+* UI Configs
+* Cluster auth preference
+* Join tokens for roles Proxy, Auth
+
+### Features we will depreciate over time
+
+All existing Teleport features will keep working with no changes, however, with this design we plan to deprecate:
+
+* The mappings in connectors `attributes_to_roles` in favor of Access Lists integrated with SCIM and identity providers. Any grants of roles will be governed by access lists.
+* Certificate extensions with roles and traits. A new access control system will no longer rely on certificate metadata to identify what roles have been assigned to users. The only data that new access control requires is information about user identity - username. Teleport will propagate grants via backend. This will lower Teleport’s resiliency to auth server failures, but this will be compensated with modern database backends like CockroachDB that provide multi-region resiliency and failover. 
+* Label matchers and resource matchers in roles. We will have to support those for a long time, but those labels will always apply at the granted scope to resources in a resource group, and will become redundant, with later versions of Teleport relying on auto-discovery and assignment of resources to resource groups.
+* Login rules. We will recommend replacing login rules with access lists that provide similar functionality.
+
+## User stories
+
+Let’s get back to the issues we outlined  in the start of this RFD and review how the new system will  help to resolve them.
+
+### Gradual scoping
+
+Most Teleport customers will start with all resources and roles in the cluster scope. 
+
+We will let them introduce resource groups gradually. Let’s create two resource groups, `prod` and `dev` with resource groups `west` and `lab`.
+
+Here is the resource groups hierarchy, where we will assume that mars and luna servers matched the assignments:
+
+```mermaid
+graph TD;
+    luna(Server luna)-->west;
+    west-->prod;
+    prod-->cluster;
+    lab-->dev;
+    mars(Server mars)-->lab;
+    dev-->cluster;
+```
+
+```yaml
+kind: resource_group
+metadata:
+  name: prod
+---
+kind: resource_group
+metadata:
+  name: west
+  parent_resource_group: prod
+  # would be nice if we could match on AWS specific right away with match_aws
```

Comment:
We can. AWS imported labels are prefixed with `aws` like `aws/account-id` so it makes sense to support them.